File: regions/system/library/go/building-blocks/registry.go
package buildingblocks

import (
	"context"
	"fmt"
	"sync"
)

// ComponentRegistry はビルディングブロックコンポーネントの登録・管理を行うレジストリ。
// スレッドセーフな RWMutex を用いて並行アクセスに対応する。
type ComponentRegistry struct {
	mu         sync.RWMutex
	components map[string]Component
}

// NewComponentRegistry は新しい ComponentRegistry を生成して返す。
func NewComponentRegistry() *ComponentRegistry {
	return &ComponentRegistry{
		components: make(map[string]Component),
	}
}

// Register はコンポーネントをレジストリに登録する。
// 同名のコンポーネントが既に存在する場合はエラーを返す。
func (r *ComponentRegistry) Register(component Component) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	name := component.Name()
	if _, exists := r.components[name]; exists {
		return fmt.Errorf("コンポーネント '%s' は既に登録されています", name)
	}
	r.components[name] = component
	return nil
}

// Get は名前でコンポーネントを取得する。存在しない場合は nil, false を返す。
func (r *ComponentRegistry) Get(name string) (Component, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()

	c, ok := r.components[name]
	return c, ok
}

// InitAll は登録済みの全コンポーネントを初期化する。
// いずれかのコンポーネントの初期化に失敗した場合、その時点でエラーを返す。
func (r *ComponentRegistry) InitAll(ctx context.Context) error {
	r.mu.RLock()
	defer r.mu.RUnlock()

	for name, component := range r.components {
		meta := Metadata{Name: name, Version: component.Version()}
		if err := component.Init(ctx, meta); err != nil {
			return fmt.Errorf("コンポーネント '%s' の初期化に失敗しました: %w", name, err)
		}
	}
	return nil
}

// CloseAll は登録済みの全コンポーネントをクローズする。
// いずれかのコンポーネントのクローズに失敗した場合、その時点でエラーを返す。
func (r *ComponentRegistry) CloseAll(ctx context.Context) error {
	r.mu.RLock()
	defer r.mu.RUnlock()

	for name, component := range r.components {
		if err := component.Close(ctx); err != nil {
			return fmt.Errorf("コンポーネント '%s' のクローズに失敗しました: %w", name, err)
		}
	}
	return nil
}

// StatusAll は登録済みの全コンポーネントのステータスをマップとして返す。
func (r *ComponentRegistry) StatusAll(ctx context.Context) map[string]ComponentStatus {
	r.mu.RLock()
	defer r.mu.RUnlock()

	statuses := make(map[string]ComponentStatus, len(r.components))
	for name, component := range r.components {
		statuses[name] = component.Status(ctx)
	}
	return statuses
}

File: regions/system/library/typescript/building-blocks/src/registry.ts
// ComponentRegistry はビルディングブロックコンポーネントの登録・管理を行うクラス。
// Map を用いてコンポーネントを名前で管理し、一括初期化・クローズ・ステータス取得をサポートする。

import type { Component, ComponentStatus } from './component.js';

export class ComponentRegistry {
  // 登録されたコンポーネントを名前をキーとして管理するマップ
  private readonly components = new Map<string, Component>();

  // コンポーネントをレジストリに登録する。同名が既に存在する場合はエラーをスローする。
  register(component: Component): void {
    if (this.components.has(component.name)) {
      throw new Error(`コンポーネント '${component.name}' は既に登録されています`);
    }
    this.components.set(component.name, component);
  }

  // 名前でコンポーネントを取得する。存在しない場合は undefined を返す。
  get(name: string): Component | undefined {
    return this.components.get(name);
  }

  // 登録済みの全コンポーネントを順次初期化する。
  // いずれかの初期化に失敗した場合、その時点でエラーをスローする。
  async initAll(): Promise<void> {
    for (const [name, component] of this.components) {
      try {
        await component.init();
      } catch (err) {
        throw new Error(`コンポーネント '${name}' の初期化に失敗しました: ${err}`);
      }
    }
  }

  // 登録済みの全コンポーネントを順次クローズする。
  // いずれかのクローズに失敗した場合、その時点でエラーをスローする。
  async closeAll(): Promise<void> {
    for (const [name, component] of this.components) {
      try {
        await component.close();
      } catch (err) {
        throw new Error(`コンポーネント '${name}' のクローズに失敗しました: ${err}`);
      }
    }
  }

  // 登録済みの全コンポーネントのステータスをオブジェクトとして返す。
  async statusAll(): Promise<Record<string, ComponentStatus>> {
    const statuses: Record<string, ComponentStatus> = {};
    for (const [name, component] of this.components) {
      statuses[name] = await component.status();
    }
    return statuses;
  }
}
